### PR TITLE
[SYCL] Fix warnings in sycl/source Part 2

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**

### CODE REVIEW

File: sycl/source/detail/kernel_program_cache.hpp [hunk 1]
```diff
@@ -442,6 +442,7 @@ class KernelProgramCache {
         return;
 
       // Save reference between the program and the fast cache key.
+      std::unique_lock<std::mutex> Lock(MKernelFastCacheMutex);
```
Comment (aelovikov-intel) [inline sycl/source/detail/kernel_program_cache.hpp:445]: What warning detected this? Should it go into a separate PR as a possible race condition fix?

File: sycl/source/detail/kernel_program_cache.hpp [hunk 2]
```diff
@@ -617,13 +620,20 @@ class KernelProgramCache {
...
+    {
+      std::lock_guard<std::mutex> L2(MKernelsPerProgramCacheMutex);
+      MKernelsPerProgramCache = KernelCacheT{};
+    }
+
+    {
+      std::lock_guard<std::mutex> L3(MKernelFastCacheMutex);
+      MKernelFastCache = KernelFastCacheT{};
+      MProgramToKernelFastCacheKeyMap.clear();
+    }
```
Comment (aelovikov-intel) [inline sycl/source/detail/kernel_program_cache.hpp:636]: Are you sure the intent wasn't that update to all of them should be one atomic operation?

**Review by aelovikov-intel — APPROVED** [review]

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/source/detail/kernel_program_cache.hpp (hunk 1 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/kernel_program_cache.hpp:445]: warning was about race condition, author left a comment that the container should be protected by this mutex forgot to add it here.

Comment (KseniyaTikhomirova) [reply sycl/source/detail/kernel_program_cache.hpp:445]: hm... I searched a code and found production place with reset call. It means than comment is not relevant any more. I will rewrite this part again. Thanks for raised concern!

Comment (KseniyaTikhomirova) [reply sycl/source/detail/kernel_program_cache.hpp:445]: fixed, restored atomicity of original code and moved eviction mutex to the top. in some code places it is used as "parent" mutex so now order of mutex locks should be aligned and not causing dead lock.

**Comment by KseniyaTikhomirova** [comment]
removing +1 to get a new review

File: sycl/source/detail/graph_impl.cpp [hunk 3]
```diff
@@ -1639,7 +1639,7 @@ node modifiable_command_graph::addImpl(dynamic_command_group &DynCGF,
 
   graph_impl::WriteLock Lock(impl->MMutex);
   std::shared_ptr<detail::node_impl> NodeImpl = impl->add(DynCGFImpl, DepImpls);
-  return sycl::detail::createSyclObjFromImpl<node>(NodeImpl);
+  return sycl::detail::createSyclObjFromImpl<node>(std::move(NodeImpl));
```
Comment (konradkusiak97) [inline sycl/source/detail/graph_impl.cpp:1642]: Is this of any benefit if `createSyclObjFromImpl()` doesn't accept r-value reference?

File: sycl/source/detail/graph_impl.cpp [hunk 4]
```diff
@@ -1989,8 +1989,8 @@ void dynamic_command_group_impl::finalizeCGFList(
     // shared_ptr<detail::CGExecKernel> to store
     sycl::detail::CG *RawCGPtr = Handler.impl->MGraphNodeCG.release();
     auto RawCGExecPtr = static_cast<sycl::detail::CGExecKernel *>(RawCGPtr);
-    auto CGExecSP = std::shared_ptr<sycl::detail::CGExecKernel>(RawCGExecPtr);
-    MKernels.push_back(CGExecSP);
+    MKernels.push_back(
+        std::shared_ptr<sycl::detail::CGExecKernel>(RawCGExecPtr));
```
Comment (konradkusiak97) [inline sycl/source/detail/graph_impl.cpp:1993]: Was this silencing a warning or is it just a nit?

**Review by konradkusiak97 — APPROVED** [review]
LGTM

File: sycl/source/detail/graph_impl.cpp (hunk 3 above)
Comment (konradkusiak97) [reply sycl/source/detail/graph_impl.cpp:1642]: A right, I guess it removes one of the copies - will call a move constructor of shared_ptr.

File: sycl/source/detail/graph_impl.cpp (hunk 4 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/graph_impl.cpp:1993]: it also removes extra copy

**Comment by KseniyaTikhomirova** [comment]
@intel/llvm-gatekeepers this PR is ready to be merged, thanks.
